src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Usage/ImplementGenericMathInterfacesCorrectlyTests.cs

```diff
@@ -0,0 +1,315 @@
+    {
+        throw new NotImplementedException();
+    }
+
+    public static bool TryParse(string s, IFormatProvider provider, out CorrectImplementation result)
+    {
+        throw new NotImplementedException();
+    }
+}
+", VerifyCS.Diagnostic(ImplementGenericMathInterfacesCorrectly.GMInterfacesRule).WithLocation(0).WithArguments("IParsable", "WrongImplementation", "TMe")).RunAsync();
+        }
+
+        [Fact]
+        public async Task DerivedUsedBaseTypeaAsTypeParameters()
+        {
+            await PopulateTestCs(@"
+using System;
+using System.Numerics;
+
+class Base : IAdditionOperators<Base, int, int>
+{
+    public static int operator +(Base left, int right)
+    {
+        throw new NotImplementedException();
+    }
+
+    public static int operator checked +(Base left, int right)
+    {
+        throw new NotImplementedException();
+    }
+}
+
+class Derived : Base, IAdditionOperators<{|#0:Base|}, int, int> // 'IAdditionOperators' interface requires the derived type 'Derived' used for the 'TSelf' type parameter
+{
+    static int IAdditionOperators<Base, int, int>.operator +(Base left, int right)
+    {
+        throw new NotImplementedException();
+    }
+
+    static int IAdditionOperators<Base, int, int>.operator checked +(Base left, int right)
+    {
+        throw new NotImplementedException();
+    }
+}
+
+class DerivedCorrect : Base, IAdditionOperators<DerivedCorrect, int, int>
+{
+    public static int operator +(DerivedCorrect left, int right)
+    {
+        throw new NotImplementedException();
+    }
+}
+", VerifyCS.Diagnostic(ImplementGenericMathInterfacesCorrectly.GMInterfacesRule).WithLocation(0).WithArguments("IAdditionOperators", "Derived", "TSelf")).RunAsync();
+        }
+
+        [Fact]
+        public async Task BaseUsedDerivedTypeAsTypeParameter()
+        {
+            await PopulateTestCs(@"
+using System;
+using System.Numerics;
+
+class Base : IAdditionOperators<{|#0:Derived|}, int, int> // 'IAdditionOperators' interface requires the derived type 'Base' used for the 'TSelf' type parameter
+{
+    static int IAdditionOperators<Derived, int, int>.operator +(Derived left, int right)
+    {
+        throw new NotImplementedException();
+    }
+}
+
+class Derived : Base, IAdditionOperators<Derived, int, int>
+{
+    public static int operator +(Derived left, int right)
+    {
+        throw new NotImplementedException();
+    }
+}
+", VerifyCS.Diagnostic(ImplementGenericMathInterfacesCorrectly.GMInterfacesRule).WithLocation(0).WithArguments("IAdditionOperators", "Base", "TSelf")).RunAsync();
+        }
+
+        private static VerifyCS.Test PopulateTestCs(string sourceCode, params DiagnosticResult[] expected)
+        {
+            var test = new VerifyCS.Test
+            {
+                TestCode = sourceCode,
+                ReferenceAssemblies = ReferenceAssemblies.Net.Net70,
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.Preview
```

Comment:
Does this need to be `Preview`?

Reply:
The latest language version in the repo is C# 10, by my understanding the static abstract is in C#11, not sure if its still in preview though

Reply:
Yeah, when use c#10 getting error `error CS8703: The modifier 'static' is not valid for this item in C# 10.0. Please use language version 'preview' or greater.` We could ask for adding C# 11 if needed

Reply:
Hmm, that's odd. It would probably make the most sense to test with the latest Roslyn version, wouldn't it?

Reply:
I think its necessary. We don't officially support C# 11 on anything prior to .NET 7

Reply:
These tests are for interfaces we introduced in .NET 7, and line 306 explicitly says `ReferenceAssemblies = ReferenceAssemblies.Net.Net70,`. My understanding of why this is necessary is because we aren't using a recent Roslyn version in this repo. Is that true?

It looks like the version was updated with https://github.com/dotnet/roslyn-analyzers/pull/6114, so I wonder if setting this to `Preview` is still necessary with the latest main.

Reply:
Yep now we have `LanguageVersion.CSharp11`, updating